ethcore/src/snapshot/service.rs

```diff
@@ -406,55 +416,136 @@ impl Service {
-			match e.kind() {
-				ErrorKind::NotFound => {},
-				_ => return Err(e.into()),
+			// delete and restore the restoration dir.
+			if let Err(e) = fs::remove_dir_all(&rest_dir) {
+				match e.kind() {
+					ErrorKind::NotFound => {},
+					_ => return Err(e.into()),
+				}
 			}
+
+			fs::create_dir_all(&rest_dir)?;
+
+			// make new restoration.
+			let writer = match recover {
+				true => Some(LooseWriter::new(self.temp_recovery_dir())?),
+				false => None
+			};
+
+			let params = RestorationParams {
+				manifest: manifest.clone(),
+				pruning: self.pruning,
+				db: self.restoration_db_handler.open(&self.restoration_db())?,
+				writer: writer,
+				genesis: &self.genesis_block,
+				guard: Guard::new(rest_db),
+				engine: &*self.engine,
+			};
+
+			let state_chunks = manifest.state_hashes.len();
+			let block_chunks = manifest.block_hashes.len();
+
+			*res = Some(Restoration::new(params)?);
+
+			*self.status.lock() = RestorationStatus::Ongoing {
+				state_chunks: state_chunks as u32,
+				block_chunks: block_chunks as u32,
+				state_chunks_done: self.state_chunks.load(Ordering::SeqCst) as u32,
+				block_chunks_done: self.block_chunks.load(Ordering::SeqCst) as u32,
+			};
+
+			self.restoring_snapshot.store(true, Ordering::SeqCst);
+
+			// Import previous chunks, continue if it fails
+			self.import_prev_chunks(&mut res, manifest).ok();
+			self.restoration_ready.store(true, Ordering::SeqCst);
 		}
 
-		fs::create_dir_all(&rest_dir)?;
+		Ok(())
+	}
+
+	/// Import the previous chunks into the current restoration
+	fn import_prev_chunks(&self, restoration: &mut Option<Restoration>, manifest: ManifestData) -> Result<(), Error> {
+		let prev_chunks = self.prev_chunks_dir();
 
-		// make new restoration.
-		let writer = match recover {
-			true => Some(LooseWriter::new(self.temp_recovery_dir())?),
-			false => None
-		};
+		// Restore previous snapshot chunks
+		let files = fs::read_dir(prev_chunks.as_path())?;
+		let mut num_temp_chunks = 0;
 
-		let params = RestorationParams {
-			manifest: manifest,
-			pruning: self.pruning,
-			db: self.restoration_db_handler.open(&self.restoration_db())?,
-			writer: writer,
-			genesis: &self.genesis_block,
-			guard: Guard::new(rest_dir),
-			engine: &*self.engine,
-		};
+		for prev_chunk_file in files {
+			// Don't go over all the files if the restoration has been aborted
+			if !self.restoring_snapshot.load(Ordering::SeqCst) {
+				trace!(target:"snapshot", "Aborting importing previous chunks");
+				return Ok(());
+			}
+			// Import the chunk, don't fail and continue if one fails
+			match self.import_prev_chunk(restoration, &manifest, prev_chunk_file) {
+				Ok(_) => num_temp_chunks += 1,
+				Err(e) => trace!(target: "snapshot", "Error importing chunk: {:?}", e),
+			}
+		}
 
-		let state_chunks = params.manifest.state_hashes.len();
-		let block_chunks = params.manifest.block_hashes.len();
+		trace!(target:"snapshot", "Imported {} previous chunks", num_temp_chunks);
```

Comment:
Judging by the way it is computed, I think, `num_temp_chunks` is a number of all chunks in temp dir. But since it's used only for debugging, it's not that important.